src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/MintermClassifier.cs

```diff
@@ -20,81 +20,108 @@ namespace System.Text.RegularExpressions.Symbolic
+            _isAsciiOnly = true;
+            for (int mintermId = 1; mintermId < minterms.Length; mintermId++)
             {
-                // Each supplied minterm BDD decides whether a given character maps to it or not.
-                // We need to combine all of those into a multi-terminal BDD that decides which
-                // minterm a character maps to.  To do that, we take each minterm BDD and replace
-                // its True result with the ID of the minterm, such that a character that would
-                // have returned True for that BDD now returns the minterm ID.
-                BDD charToTargetMintermId = solver.ReplaceTrue(minterms[i], i);
-
-                // Now union this BDD with the multi-terminal BDD we've built up thus far. Unioning
-                // is valid because every character belongs to exactly one minterm and thus will
-                // only map to an ID instead of False in exactly one of the input BDDs.
-                anyCharacterToMintermId = solver.Or(anyCharacterToMintermId, charToTargetMintermId);
+                (uint, uint)[] mintermRanges = BDDRangeConverter.ToRanges(minterms[mintermId]);
+                if (mintermRanges[^1].Item2 >= 128)
+                {
+                    _isAsciiOnly = false;
+                }
             }
 
-            // Now that we have our mapping that supports any input character, we want to optimize for
-            // ASCII inputs.  Rather than forcing every input ASCII character to consult the BDD at match
-            // time, we precompute a lookup table, where each ASCII character can be used to index into the
-            // array to determine the ID for its corresponding minterm.
-            var ascii = new int[128];
-            for (int i = 0; i < ascii.Length; i++)
+            // i have never seen a regex use over 80 minterms not to speak of 255,
+            // but it's there as a fallback mechanism
+            if (minterms.Length > 255)
             {
-                ascii[i] = anyCharacterToMintermId.Find(i);
+                // over 255 unique sets also means it's never ascii only
+                int[] lookup = new int[ushort.MaxValue + 1];
+                for (int mintermId = 1; mintermId < minterms.Length; mintermId++)
+                {
+                    // precompute all assigned minterm categories
+                    (uint, uint)[] mintermRanges = BDDRangeConverter.ToRanges(minterms[mintermId]);
+                    foreach ((uint start, uint end) in mintermRanges)
+                    {
+                        // assign character ranges in bulk
+                        Span<int> slice = lookup.AsSpan((int)start, (int)(end + 1 - start));
+                        slice.Fill(mintermId);
+                    }
+                }
+                _intLookup = lookup;
+            }
+            else
+            {
+                byte[] lookup = new byte[_isAsciiOnly ? 128 : ushort.MaxValue + 1];
```

Comment:
Some extra startup cost with nonbacktracking is unavoidable, but in most cases i think the performance gains far outweigh the memory costs. It needs some comparison to see just how much it adds. For small regexes this 64K makes the bulk of the overall memory consumption. For more complex regexes this is small compared to the state space itself.  If it's regularly used and thrown away then array pooling and reclaiming could potentially make it cheap. Allocating only up to the Kelvin sign would make this about 8.5K, another option is of course adding if-else checks for these characters specifically but i'm trying to avoid any extra overhead at the very center of the hot loop as it'll immediately show.